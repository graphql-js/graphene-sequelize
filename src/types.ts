import { attributesToFields, associationsToFields, Fields } from "./converter";
import { setSequelizeModel } from "./reflection";
import { assertSequelizeModel, Model, assertRegistry } from "./utils";
import { Registry, getGlobalRegistry } from "./registry";
import { getFields, ObjectTypeConfig, ObjectType } from "graphene-js";
import { ModelAttributes, ModelAssociations } from "./sequelize";

export type SequelizeObjectTypeConfig = ObjectTypeConfig & {
  model: Model;
  registry?: Registry;
};

export const SequelizeObjectType = (opts: SequelizeObjectTypeConfig) => <
  T extends { new (...args: any[]): {}; [key: string]: any }
>(
  target: T
): T => {
  let { model, registry, ...objectTypeOpts } = opts;
  assertSequelizeModel(model);
  if (!registry) {
    registry = getGlobalRegistry();
  } else {
    assertRegistry(registry);
  }
  let attributes: ModelAttributes = (model as any).attributes;
  let associations: ModelAssociations = (model as any).associations;
  // console.log((model as any).associations);
  // console.log(target.name, model.);
  let modelAttributeFields: Fields = attributesToFields(attributes);
  let modelAssociationFields: Fields = associationsToFields(
    associations,
    registry
  );
  let baseFields = getFields(target);
  let modelFields: Fields = {
    ...modelAttributeFields,
    ...modelAssociationFields
  };
  for (let fieldName in modelFields) {
    if (fieldName in baseFields) {
      // If the field is already defined on the base
      // objecttype, then we just skip it to avoid conflict.
      continue;
    }
    let field = modelFields[fieldName];
    // We setup the field
    field(target.prototype, fieldName);
  }
  setSequelizeModel(target, model);
  registry.register(target);
  return ObjectType(objectTypeOpts)(target);
};
