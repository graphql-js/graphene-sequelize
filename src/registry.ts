import { assertSequelizeModel, Model } from "./utils";
import * as Sequelize from "sequelize";
import { getSequelizeModel } from "./reflection";

export type RegistryMapping = Sequelize.Model<any, any>[];

export class Registry {
  public mapping: RegistryMapping[];
  constructor() {
    this.mapping = [];
  }
  public register(sequelizeType: any) {
    let model: Model = getSequelizeModel(sequelizeType);
    this.mapping.push(sequelizeType);
  }
  public getTypeForModel(model: Sequelize.Model<any, any>) {
    assertSequelizeModel(model);
    for (let sequelizeType of this.mapping) {
      let sequelizeTypeModel: Model = getSequelizeModel(sequelizeType);
      if (sequelizeTypeModel === model) {
        return sequelizeType;
      }
    }
    return null;
  }
}

let registry: Registry | null = null;

export const getGlobalRegistry = (): Registry => {
  if (registry === null) {
    registry = new Registry();
  }
  return registry;
};

export const resetGlobalRegistry = () => {
  registry = null;
};
