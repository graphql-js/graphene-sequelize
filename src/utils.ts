import * as Sequelize from "sequelize";
import { Registry } from "./registry";

export type Model = Sequelize.Model<any, any>;

export const assertSequelizeModel = (model: any) => {
  if (model instanceof <any>Sequelize.Model) {
    throw new Error(
      `Expected to receive a Sequelize.Model. Received a sequelize instance instead: ${model}.`
    );
  }
  if (!model || model.constructor !== Sequelize.Model.constructor) {
    throw new Error(
      `Expected to receive a Sequelize.Model. Received ${model} instead.`
    );
  }
};

export const assertRegistry = (registry: any) => {
  if (!(registry instanceof Registry)) {
    throw new Error(
      `Expected to receive a Registry instance. Received ${registry} instead.`
    );
  }
};
