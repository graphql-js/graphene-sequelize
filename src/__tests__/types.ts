import { GraphQLObjectType } from "graphql";
import * as Sequelize from "sequelize";
import {
  getFields,
  getGraphQLType,
  ObjectType,
  Field,
  ID,
  Schema
} from "graphene-js";

import { Model } from "./../utils";
import { SequelizeObjectType } from "./../types";
import { Registry } from "./../registry";
import { resetGlobalRegistry } from "..";

describe("utils", () => {
  const sequelize = new Sequelize("db", "username", "pass", {
    dialect: "sqlite",
    storage: ":memory:",
    logging: () => null
  });

  let userModel: Model = sequelize.define("user", {
    name: Sequelize.STRING,
    lastName: { type: Sequelize.STRING, comment: "The last name" }
  });

  let companyModel: Model = sequelize.define("company", {
    name: Sequelize.STRING
  });

  // companyModel.hasOne(userModel, { as: "owner" });
  companyModel.belongsToMany(userModel, {
    as: "employees",
    through: "CompanyEmployees",
    foreignKey: "companyId"
  });

  const createCompany = async (employees: any[]) => {
    let company = await companyModel.create({
      name: "GraphQL"
    });
    await company.setEmployees(employees);
    return company;
  };

  const createUsers = async () => {
    await userModel.create({
      name: "Lee",
      lastName: "Byron"
    });

    await userModel.create({
      name: "Oleg",
      lastName: "Ilyenko"
    });

    await userModel.create({
      name: "Sashko",
      lastName: "Stubailo"
    });

    await userModel.create({
      name: "Johannes",
      lastName: "Schickling"
    });
  };
  beforeEach(async () => {
    resetGlobalRegistry();
    await sequelize.sync();
  });

  afterEach(async () => {
    await sequelize.drop();
  });

  test("SequelizeObjectType works", () => {
    @SequelizeObjectType({ model: userModel })
    class User {}

    expect(
      (<GraphQLObjectType>getGraphQLType(User)).getFields()
    ).toMatchSnapshot();
  });

  test("SequelizeObjectType can overwrite field", () => {
    @SequelizeObjectType({ model: userModel })
    class User {
      @Field(String) public id: string;
    }

    expect(
      (<GraphQLObjectType>getGraphQLType(User)).getFields()
    ).toMatchSnapshot();
  });

  test("SequelizeObjectType works", () => {
    expect(() => {
      @SequelizeObjectType({ model: <any>null })
      class User {}
    }).toThrowErrorMatchingSnapshot();
  });

  test("SequelizeObjectType works", () => {
    expect(() => {
      @SequelizeObjectType({ model: userModel, registry: <any>{} })
      class User {}
    }).toThrowErrorMatchingSnapshot();
  });

  test("SequelizeObjectType can query", async () => {
    @SequelizeObjectType({ model: userModel })
    class User {
      @Field(String)
      private fullName(this: any) {
        return `${this.name} ${this.lastName}`;
      }
    }

    @ObjectType()
    class Query {
      @Field(User, { args: { id: ID } })
      private getUser({ id }: { id: string }) {
        return userModel.findById(id);
      }

      @Field([User])
      private allUsers() {
        return userModel.findAll();
      }
    }

    await createUsers();

    const schema = new Schema({ query: Query });
    const result = await schema.execute(
      `{ allUsers { id name lastName fullName } }`
    );
    expect(result).toMatchSnapshot();
  });

  test("SequelizeObjectType set many relations", () => {
    @SequelizeObjectType({ model: userModel })
    class User {}

    @SequelizeObjectType({ model: companyModel })
    class Company {}

    expect(
      (<GraphQLObjectType>getGraphQLType(Company)).getFields()
    ).toHaveProperty("employees");

    expect(
      (<GraphQLObjectType>getGraphQLType(Company)).getFields()
    ).toMatchSnapshot();
  });

  test("SequelizeObjectType set many relations (reverse registration)", () => {
    @SequelizeObjectType({ model: companyModel })
    class Company {}

    @SequelizeObjectType({ model: userModel })
    class User {}

    expect(
      (<GraphQLObjectType>getGraphQLType(Company)).getFields()
    ).toHaveProperty("employees");

    expect(
      (<GraphQLObjectType>getGraphQLType(Company)).getFields()
    ).toMatchSnapshot();
  });

  test("SequelizeObjectType many relations can resolve", async () => {
    @SequelizeObjectType({ model: companyModel })
    class Company {}

    @SequelizeObjectType({ model: userModel })
    class User {
      @Field(String)
      private fullName(this: any) {
        return `${this.name} ${this.lastName}`;
      }
    }

    await createUsers();
    let employees = await userModel.findAll();
    let company = await createCompany(employees);
    @ObjectType()
    class Query {
      @Field(Company)
      protected company() {
        return company;
      }
    }

    const schema = new Schema({ query: Query });
    const result = await schema.execute(
      `{ company { id, employees { id fullName } } }`
    );
    expect(result).toMatchSnapshot();
  });
});
