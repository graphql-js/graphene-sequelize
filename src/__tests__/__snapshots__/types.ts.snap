// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils SequelizeObjectType can overwrite field 1`] = `
Object {
  "createdAt": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "createdAt",
    "resolve": [Function],
    "type": "DateTime!",
  },
  "id": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "id",
    "resolve": [Function],
    "type": "String",
  },
  "lastName": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": "The last name",
    "isDeprecated": false,
    "name": "lastName",
    "resolve": [Function],
    "type": "String",
  },
  "name": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "name",
    "resolve": [Function],
    "type": "String",
  },
  "updatedAt": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "updatedAt",
    "resolve": [Function],
    "type": "DateTime!",
  },
}
`;

exports[`utils SequelizeObjectType can query 1`] = `
Object {
  "data": Object {
    "allUsers": Array [
      Object {
        "fullName": "Lee Byron",
        "id": 1,
        "lastName": "Byron",
        "name": "Lee",
      },
      Object {
        "fullName": "Oleg Ilyenko",
        "id": 2,
        "lastName": "Ilyenko",
        "name": "Oleg",
      },
      Object {
        "fullName": "Sashko Stubailo",
        "id": 3,
        "lastName": "Stubailo",
        "name": "Sashko",
      },
      Object {
        "fullName": "Johannes Schickling",
        "id": 4,
        "lastName": "Schickling",
        "name": "Johannes",
      },
    ],
  },
}
`;

exports[`utils SequelizeObjectType many relations can resolve 1`] = `
Object {
  "data": Object {
    "company": Object {
      "employees": Array [
        Object {
          "fullName": "Lee Byron",
          "id": 1,
        },
        Object {
          "fullName": "Oleg Ilyenko",
          "id": 2,
        },
        Object {
          "fullName": "Sashko Stubailo",
          "id": 3,
        },
        Object {
          "fullName": "Johannes Schickling",
          "id": 4,
        },
      ],
      "id": 1,
    },
  },
}
`;

exports[`utils SequelizeObjectType set many relations (reverse registration) 1`] = `
Object {
  "createdAt": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "createdAt",
    "resolve": [Function],
    "type": "DateTime!",
  },
  "employees": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "employees",
    "resolve": [Function],
    "type": "[User]",
  },
  "id": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "id",
    "resolve": [Function],
    "type": "Int!",
  },
  "name": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "name",
    "resolve": [Function],
    "type": "String",
  },
  "updatedAt": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "updatedAt",
    "resolve": [Function],
    "type": "DateTime!",
  },
}
`;

exports[`utils SequelizeObjectType set many relations 1`] = `
Object {
  "createdAt": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "createdAt",
    "resolve": [Function],
    "type": "DateTime!",
  },
  "employees": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "employees",
    "resolve": [Function],
    "type": "[User]",
  },
  "id": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "id",
    "resolve": [Function],
    "type": "Int!",
  },
  "name": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "name",
    "resolve": [Function],
    "type": "String",
  },
  "updatedAt": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "updatedAt",
    "resolve": [Function],
    "type": "DateTime!",
  },
}
`;

exports[`utils SequelizeObjectType works 1`] = `
Object {
  "createdAt": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "createdAt",
    "resolve": [Function],
    "type": "DateTime!",
  },
  "id": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "id",
    "resolve": [Function],
    "type": "Int!",
  },
  "lastName": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": "The last name",
    "isDeprecated": false,
    "name": "lastName",
    "resolve": [Function],
    "type": "String",
  },
  "name": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "name",
    "resolve": [Function],
    "type": "String",
  },
  "updatedAt": Object {
    "args": Array [],
    "deprecationReason": undefined,
    "description": undefined,
    "isDeprecated": false,
    "name": "updatedAt",
    "resolve": [Function],
    "type": "DateTime!",
  },
}
`;

exports[`utils SequelizeObjectType works 2`] = `"Expected to receive a Sequelize.Model. Received null instead."`;

exports[`utils SequelizeObjectType works 3`] = `"Expected to receive a Registry instance. Received [object Object] instead."`;
